symmetries.py

```diff
@@ -128,3 +129,49 @@ def rotate(tensor):
             axis=0)
 
     return rotate(x), pi_rot
+
+
+def rotate_train_nchw(x, pi):
+    sym = tf.random_uniform(
+        [],
+        minval=0,
+        maxval=len(SYMMETRIES),
+        dtype=tf.int32,
+        seed=123)
+
+    def rotate(tensor):
+        # flipLeftRight
+        tensor = tf.where(
+            tf.bitwise.bitwise_and(sym, 1) > 0,
+            tf.reverse(tensor, axis=[1]),
+            tensor)
+        # flipUpDown
+        tensor = tf.where(
+            tf.bitwise.bitwise_and(sym, 2) > 0,
+            tf.reverse(tensor, axis=[2]),
+            tensor)
+        # flipDiagonal
+        tensor = tf.where(
+            tf.bitwise.bitwise_and(sym, 4) > 0,
+            tf.transpose(tensor, perm=[0, 2, 1]),
+            tensor)
+        return tensor
+
+    # TODO(tommadams): use tf.ensure_shape instead of tf.assert_equal.
+    squares = go.N * go.N
+    assert_shape_pi = tf.assert_equal(pi.shape.as_list(), [squares + 1])
+
+    x_shape = x.shape.as_list()
+    assert_shape_x = tf.assert_equal(x_shape, [x_shape[0], go.N, go.N])
+
+    pi_move = tf.slice(pi, [0], [squares], name="slice_moves")
+    pi_pass = tf.slice(pi, [squares], [1], name="slice_pass")
+    # Add a final dim so that x and pi have same shape: [N,N,num_features].
+    pi_n_by_n = tf.reshape(pi_move, [1, go.N, go.N])
+
+    with tf.control_dependencies([assert_shape_x, assert_shape_pi]):
+        pi_rot = tf.concat(
+            [tf.reshape(rotate(pi_n_by_n), [squares]), pi_pass],
+            axis=0)
```

Comment:
consider passing in layout to rotate then not having to repeat this block.
I'm not a big fan but maybe.

Reply:
Yeah, I spent a while trying to figure out the cleanest way to do this and then just punted and went with the golang approach: copy-paste.